api/types/entrySchema.go

```diff
@@ -47,49 +100,80 @@ func (s *EntrySchema) UnmarshalJSON(data []byte) error {
 	if rawGraph.Size() <= 0 {
 		return fmt.Errorf("expected a non-empty JSON object but got %v instead", string(data))
 	}
-	s.graph = linkedhashmap.New()
+	graph := linkedhashmap.New()
 
-	// Convert each node in the rawGraph to an *EntrySchema
-	// object
+	// Convert each node in the rawGraph to a plugin.EntrySchema
+	// object. This is also where we validate the data.
 	var err error
-	firstElem := true
 	rawGraph.Each(func(key interface{}, value interface{}) {
 		if err != nil {
 			return
 		}
-		var node *EntrySchema
-		if firstElem {
-			node = s
-			firstElem = false
-		} else {
-			node = &EntrySchema{
-				graph: s.graph,
-			}
-		}
-		node.typeID = key.(string)
-		err = deepcopy.Copy(&node.EntrySchema, value.(map[string]interface{}))
+		var schema plugin.EntrySchema
+		err = deepcopy.Copy(&schema, value.(map[string]interface{}))
 		if err != nil {
 			return
 		}
-		s.graph.Put(key, node)
+		if len(schema.Label) <= 0 {
+			err = fmt.Errorf("label for %v was not provided", key.(string))
+			return
+		}
+		graph.Put(key, schema)
 	})
 	if err != nil {
 		return err
 	}
 
-	// Fill-in the children map
-	s.graph.Each(func(key interface{}, value interface{}) {
-		schema := value.(*EntrySchema)
-		for _, childTypeID := range schema.EntrySchema.Children {
-			rawChild, _ := s.graph.Get(childTypeID)
-			schema.children = append(schema.children, rawChild.(*EntrySchema))
+	// Now fill-in the stree.
+	var fillStree func(string, string, map[string]*EntrySchema) *EntrySchema
+	fillStree = func(path string, typeID string, visited map[string]*EntrySchema) *EntrySchema {
+		if node, ok := visited[typeID]; ok {
+			return node
 		}
-	})
+		schema, _ := graph.Get(typeID)
+		node := &EntrySchema{
+			EntrySchema: schema.(plugin.EntrySchema),
+			typeID:      typeID,
+			path:        path,
+		}
+		if len(path) <= 0 {
+			// This is the root
+			node.path = node.Label()
+		} else {
+			// This is some intermediate node
+			node.path = path + "/" + node.Label()
+		}
+		visited[typeID] = node
+		for _, childTypeID := range node.EntrySchema.Children {
+			node.children = append(node.children, fillStree(node.path, childTypeID, visited))
+		}
+		delete(visited, typeID)
+		return node
+	}
+	it := graph.Iterator()
+	it.First()
+	root := fillStree("", it.Key().(string), make(map[string]*EntrySchema))
+	(*s) = (*root)
+	s.graph = graph
 
 	return nil
 }
 
-// TypeID returns the entry's type ID. This should be unique.
+// Path returns the unique path to this specific entry's schema
+// (relative to the stree root). The path consists of
+//    <root_label>/<parent1_label>/.../<label>
```

Comment:
This format makes the kind primary a bit annoying to use from the mountpoint. For example, if I want to find only Docker containers, I'll need to do something like `find -k "mountpoint/docker*container"` (I can't do `find -k "*docker*container"` because that will return containers in the `mock_docker` plugin).

I've been thinking about some alternatives:
1. Change `root_label` => `.` (which would indicate that the path is relative to the root). Then the previous example becomes something like `find -k "./docker*container"`. Problem with this is if one has multiple paths, e.g. `find <path1> <path2>`, it makes it impossible to filter on the kind of `<path1>`, `<path2>`.

2. Add some munging to the glob where if it starts with `.`, we will consider that the kind of `<path1>`, `<path2>`, depending on which paths' being walked. So, `./docker*container` would be munged to `mountpoint/docker*container`.

3. Change the mountpoint label to `.`, but keep everything else as-is. Think this makes the most sense, since the mountpoint label isn't being used anywhere.

Reply:
What about treating an absolute path as the start of the schema path? So `/docker*container`? Not sure the `.` is necessary to clarify that description.

Reply:
Yeah, could also do that too. Would that resolve the issue in (1)?

Reply:
It's not clear to me how any of them enable that. Maybe I'm getting confused by what `mountpoint` is. I assumed that would normally be whatever we mount Wash in, e.g. `mnt28139`.

Maybe the `/` is unnecessary because you're expecting a full match. For the initial example, why doesn't `docker/*container` match all (and only) containers? For example (1), could you give some concrete examples?

Reply:
`mountpoint` is plugin.Registry's label.

In the initial example, the schema paths of entries in the `docker` and the `mock_docker` plugin will start with `mountpoint/docker` and `mountpoint/mock_docker`, respectively (assuming we're invoking `find` in the mountpoint -- the mountpoint's entry is plugin.Registry). Since users should not care about the mountpoint label, they'd need to do something like `*docker*container` to match Docker containers -- `*docker` matches both `docker` and `mock_docker`.

The `<path1>`, `<path2>` stuff is a bit theoretical. Here's one example,
```
find <ec2_instance> <s3_bucket> -k 'instance' -o -k 'bucket'
```

Here you're filtering on entries of kind EC2 Instance or kind Bucket, so you'd print out `<ec2_instance>` and `<s3_bucket>`. The schema paths of `<ec2_instance>` and `<s3_bucket>` is (currently) `instance` and `bucket`, which is why this works. Munging the schema paths to `./` or `/` would lose their kinds, so the above expression would be incorrect.

Even though the above example is theoretical, I'd still like to get it right because there's no reason why you shouldn't be able to do that (all other predicates make it possible for you to print `<ec2_instance>` and `<s3_bucket>`, for example). 

I'm starting to like Option 3 more.

Reply:
Note that you can also do
```
find docker -k '*container'
```

to filter out Docker containers. We could also document that in the kind primary's documentation.

Reply:
Ok, so the problem is you're using glob somewhat unconventionally. In Linux, the `*` doesn't normally match separators and you have to provide a full path. Somewhere the `**` pattern was added to match directory trees. You're not passing any separators to the glob library, so it's matching them with a simple `*`.

My inclination would be to include separators because I think it avoids some common ambiguities. So your examples would be (omitting out the mountpoint label because we shouldn't use it)
- `docker/**/container`
- `**/instance`

The way you're configuring the glob calls, those examples have to be
- `/docker/*/container` to avoid ambiguities such as a plugin named `mock_docker` or `docker_new`, and new things like `dead_container` (if you're trying to be that specific). In practice maybe `/docker/*container` is fine.
- I have no idea how you're making `instance` work, because the glob library still needs a full match. I don't think the code you have here supports it, so you'd need to do something like `*instance` or `*/instance`.

The details about what the mountpoint should be called seem unimportant because glob works on full paths, and we should just exclude the mountpoint completely.

Reply:
Ok, I'm caught up. `find -path` is a pattern that treats separators as normal characters, so it doesn't behave the same as shell globbing. That's the behavior we're basing this new primary on.